### PR TITLE
Add CustomObject Service

### PR DESCRIPTION
**Pull request by ninalindl**
#### Summary
Service and implementation class for syncing CustomObjects were added, as well as integration and unit tests for it.
#### Description
Ability to sync CustomObjects was needed by a customer. To uniquely identify a CustomObject, the combination of container + key is needed.
#### Todo

- Tests
    - [ ] Unit 
    - [ ] Integration
- [ ] Documentation
- [ ] Add Release Notes entry.

#### Hints for Review
The relevant files/directories to review are the following:

- src/main/java/com/commercetools/sync/services/CustomObjectService.java

- src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java

- src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java

- src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java

- src/main/java/com/commercetools/sync/customobjects

### CODE REVIEW

File: src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        when(draftMock.getContainer()).thenReturn(customObjectContainer);
+        when(draftMock.getJavaType()).thenReturn(getCustomObjectJavaTypeForValue(convertToJavaType(JsonNode.class)));
+
+        final Optional<CustomObject<JsonNode>> customObjectOptional =
+            service.upsertCustomObject(draftMock).toCompletableFuture().join();
+
+        String expectedMsg = "Failed to create draft with key: '{key='%s', container='%s'}'";
+        assertAll(
+            () -> assertThat(customObjectOptional).isEmpty(),
+            () -> assertThat(errorMessages).hasOnlyOneElementSatisfying(message -> {
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java:219]: I see `hasOnlyOneElementSatisfying` as deprecated - can you use non-deprecated version?

File: src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    @AfterEach
+    void cleanup() {
+        reset(client);
+    }
+
+    private interface CustomObjectPagedQueryResult extends PagedQueryResult<CustomObject> {
+    }
+
+    @Test
+    void fetchCachedCustomObjectId_WithKeyAndContainer_ShouldFetchCustomObject() {
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java:81]: Is this unit test testing the same ting as `fetchCachedCustomObjectId_WithExistingCustomObject_ShouldFetchCustomObjectAndCache` in IT? If so, I would remove one of the tests. Recently we removed integration tests that were already covered by unit tests, so it's not good to reintroduce again the same problem (duplicate unit and integration tests). Check also other tests in this class for possible duplicates.

**Review by lojzatran — CHANGES_REQUESTED** [review]
In general this PR looks good, just some small stuff I found and some questions :)

File: src/main/java/com/commercetools/sync/services/CustomObjectService.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import javax.annotation.Nonnull;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+
+public interface CustomObjectService {
+
+    /**
+     * Given a {@code identifier} and convert the object to string as key, this method first checks if a cached
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/services/CustomObjectService.java:18]: The keyToId cached map in `BaseService` class requires `Map<String, String>`. Therefore we convert the identifier object to String by using `.toString()` so that it can act as a key in the cached map.

[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/services/CustomObjectService.java:18]: So the method is doing it or is this conversion also given? It's not clear from the comment.

File: src/main/java/com/commercetools/sync/services/CustomObjectService.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+     * map of {@link CustomObjectCompositeIdentifier#toString()} -&gt; ids contains the key. If not, it returns a
+     * completed future that contains an {@link Optional} that contains what this identifier maps to in the cache. If
```
[the remark this answers is not in the document]
Comment (ninalindl) [reply src/main/java/com/commercetools/sync/services/CustomObjectService.java]: I fixed it here: ff4343dfa47d8565efa8b08029371fd5c27c17d2

File: src/main/java/com/commercetools/sync/services/CustomObjectService.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+     * the cache doesn't contain the identifier; this method attempts to fetch the id of the identifier from the CTP
+     * project, caches it and returns a {@link CompletionStage}&lt;{@link Optional}&lt;{@link String}&gt;&gt;
+     * in which the result of it's completion could contain an
+     * {@link Optional} with the id inside of it or an empty {@link Optional} if no {@link CustomObject} was
+     * found in the CTP project with this key.
```
[the remark this answers is not in the document]
Comment (ninalindl) [reply src/main/java/com/commercetools/sync/services/CustomObjectService.java:25]: I copied it from the javadocs of the typeservice interface and made changes to the parameters. I agree that it is implementation-specific. Do you have any suggestions for a more generic one? I don't know what to write there instead.

File: src/integration-test/java/com/commercetools/sync/integration/commons/utils/CustomObjectITUtils.java [hunk 6]
```diff
@@ -2,23 +2,74 @@
...
+     *
+     * @param ctpClient represents the CTP project the types will be deleted from.
+     */
+    public static void deleteCustomObjects(@Nonnull final SphereClient ctpClient) {
+
+        final CustomObjectQuery<JsonNode> customObjectQuery = CustomObjectQuery.ofJsonNode();
+
+        ctpClient
+                .execute(customObjectQuery)
+                .thenApply(PagedQueryResult::getResults)
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/integration-test/java/com/commercetools/sync/integration/commons/utils/CustomObjectITUtils.java]: Unlike other resources, CustomObjectDeleteCommand doesn't support "delete all" features. It requires ID, key, container to instantiate. In addition I don't prefer to run it in infinite iteration. In the end I only delete those COs which has key/container used in Integration Test.
https://github.com/commercetools/commercetools-sync-java/pull/557/commits/3716a062d3c4872d1029948d6a99ead7c65a3d7d

File: src/integration-test/java/com/commercetools/sync/integration/commons/utils/CustomObjectITUtils.java [hunk 7]
```diff
@@ -2,23 +2,74 @@
...
+                        .map(customObject -> ctpClient
+                                .execute(CustomObjectDeleteCommand.of(customObject, JsonNode.class)))
+                        .map(CompletionStage::toCompletableFuture)
+                        .toArray(CompletableFuture[]::new));
+    }
+
+    /**
+     * Create a customObject in target CTP project, represented by provided {@code ctpClient}.
+     *
+     * @param ctpClient represents the CTP project the types will be deleted from.
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/integration-test/java/com/commercetools/sync/integration/commons/utils/CustomObjectITUtils.java]: https://github.com/commercetools/commercetools-sync-java/pull/557/commits/b44ea4a6267b9a9f1c71c6e8ddc7da6563b009bd

File: src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+
+class CustomObjectServiceImplIT {
+    private CustomObjectService customObjectService;
+    private static final String OLD_CUSTOM_OBJECT_KEY = "old_custom_object_key";
+    private static final String OLD_CUSTOM_OBJECT_CONTAINER = "old_custom_object_container";
+    private static final ObjectNode OLD_CUSTOM_OBJECT_VALUE =
+            JsonNodeFactory.instance.objectNode().put(
+                    "old_custom_object_attribute", "old_custom_object_value");
+
+    private static final String NEW_CUSTOM_OBJECT_ID = "new_custom_object_id";
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java]: https://github.com/commercetools/commercetools-sync-java/pull/557/commits/2eee4febffd5a3d9a93f1e0058f8fa51b28efdb9

File: src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java (hunk 1 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java]: /commits/4394d8ed3927c801344a04aaea5693d44fc8949d

File: src/main/java/com/commercetools/sync/services/CustomObjectService.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/services/CustomObjectService.java:18]: The conversion is doing in `CustomObjectServiceImpl`

File: src/main/java/com/commercetools/sync/services/CustomObjectService.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+     *
+     * @param  identifier the identifier object containing CustomObject key and container, by which a
+     *         {@link io.sphere.sdk.customobjects.CustomObject} id should be fetched from the CTP project.
+     * @return {@link CompletionStage}&lt;{@link Optional}&lt;{@link String}&gt;&gt; in which the result of its
+     *         completion could contain an {@link Optional} with the id inside of it or an empty {@link Optional} if no
+     *         {@link CustomObject} was found in the CTP project with this identifier.
+     */
+    @Nonnull
+    CompletionStage<Optional<String>> fetchCachedCustomObjectId(@Nonnull CustomObjectCompositeIdentifier identifier);
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/services/CustomObjectService.java:34]: Why? This method is fetching CustomObjectId as result.
Do you mean aligning the naming convension like `fetchCachedCustomObjectIdByCompositeIdentifier`?

File: src/main/java/com/commercetools/sync/customobjects/helpers/CustomObjectCompositeIdentifier.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        return this.key;
+    }
+
+    public String getContainer() {
+        return this.container;
+    }
+
+    @Override
+    public String toString() {
+        if (StringUtils.isEmpty(key) && StringUtils.isEmpty(container)) {
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/customobjects/helpers/CustomObjectCompositeIdentifier.java]: They both are not supposed to be empty.
Remove one unit test case as well.
https://github.com/commercetools/commercetools-sync-java/pull/557/commits/9c75723d0b57cfa76172af48f2a7ca84eebf53a3

File: src/main/java/com/commercetools/sync/services/CustomObjectService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/services/CustomObjectService.java:34]: Yep, I mean the naming. Now looking at it, I think it's enough to have `fetchCachedCustomObjectId`? By what is already seen in the param.

File: src/main/java/com/commercetools/sync/services/CustomObjectService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/services/CustomObjectService.java:25]:
```
    /**
     * Given an {@code identifier}, this method first checks if {@code identifier#toString()}
     * is contained in a cached map of {@link CustomObjectCompositeIdentifier#toString()} -&gt; ids . 
     * If it contains, it returns a {@link CompletionStage}&lt;{@link Optional}&lt;{@link String}&gt;&gt; 
     * in which String is what this identifier maps to in the cache. If the cache doesn't contain the identifier,
     * this method attempts to fetch the id of the identifier from the CTP project, caches it and returns
     * a {@link CompletionStage}&lt;{@link Optional}&lt;{@link String}&gt;&gt;
     * in which the {@link Optional} could contain the id inside of it.
     *
     * @param  identifier the identifier object containing CustomObject key and container, by which a
     *         {@link io.sphere.sdk.customobjects.CustomObject} id should be fetched from the CTP project.
     * @return {@link CompletionStage}&lt;{@link Optional}&lt;{@link String}&gt;&gt; in which the result of its
     *         completion could contain an {@link Optional} with the id inside of it or an empty {@link Optional} if no
     *         {@link CustomObject} was found in the CTP project with this identifier.
     */
```

File: src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+    @Test
+    void fetchCustomObject_WithExistingCustomObjectKeyAndContainer_ShouldFetchCustomObject() {
+        final Optional<CustomObject<JsonNode>> customObjectOptional = CTP_TARGET_CLIENT
+            .execute(CustomObjectQuery.ofJsonNode()
+                  .withPredicates(customObjectQueryModel ->
+                          customObjectQueryModel.key().is(OLD_CUSTOM_OBJECT_KEY).and(
+                          customObjectQueryModel.container().is(OLD_CUSTOM_OBJECT_CONTAINER))))
+            .toCompletableFuture().join().head();
+        assertThat(customObjectOptional).isNotNull();
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java:212]: Added the checking in `fetchCachedCustomObjectId_WithExistingCustomObject_ShouldFetchCustomObjectAndCache`
https://github.com/commercetools/commercetools-sync-java/pull/557/commits/44484c75c2b1f946dfc204887a41b821fffc2eee

File: src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+        final Optional<String> customObject = customObjectService.fetchCachedCustomObjectId(compositeIdentifier)
+                                                   .toCompletableFuture()
+                                                   .join();
+        assertThat(customObject).isEmpty();
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+    }
+
+    @Test
+    void fetchCachedCustomObjectId_WithExistingCustomObject_ShouldFetchCustomObjectAndCache() {
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java:104]: I don't think it can be verified. Integration test is not in same package as `BaseService` class. No way to obtain it.

File: src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java (hunk 2 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java:81]: https://github.com/commercetools/commercetools-sync-java/pull/557/commits/de95f359c568dbd9c08336dce2470caafa342faa

File: src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java:104]: Then just remove `AndCache` as I would expect that you will test it here.

File: src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java (hunk 2 above)
Comment (lojzatran) [reply src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java:81]: Can you also add a reason for @Disabled?

File: src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+        assertThat(customObjectOptional).isNotNull();
+
+        final Optional<CustomObject<JsonNode>> fetchedCustomObjectOptional =
+            executeBlocking(customObjectService.fetchCustomObject(CustomObjectCompositeIdentifier.of(
+                    OLD_CUSTOM_OBJECT_KEY, OLD_CUSTOM_OBJECT_CONTAINER)));
+        assertThat(customObjectOptional).isEqualTo(fetchedCustomObjectOptional);
+    }
+
+    @Test
+    void fetchCustomObject_WithBlankKey_ShouldNotFetchCustomObject() {
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java]: Add checking before doing fetching, also added unit test case to verify in commits
https://github.com/commercetools/commercetools-sync-java/pull/557/commits/da4d6e9b71a4a85f83e34b2b996a98a74261cce2
https://github.com/commercetools/commercetools-sync-java/pull/557/commits/84d92be4159f65fa24519e13664c072b2a406b61

File: src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java (hunk 2 above)
Comment (leungkinghin) [reply src/test/java/com/commercetools/sync/services/impl/CustomObjectServiceImplTest.java:82]: I removed the whole method already

File: src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java]: Anyway this test case has been moved from IT to Unit test

File: src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+            String key = identifier.getKey();
+            String container = identifier.getContainer();
+            if (firstAttempt) {
+                queryPredicate = queryModel.container().is(container).and(queryModel.key().is(key));
+                firstAttempt = false;
+            } else {
+                queryPredicate = queryPredicate.or(queryModel.container().is(container).and(queryModel.key().is(key)));
+            }
+        }
+        return queryPredicate;
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java:107]: Fixed.
https://github.com/commercetools/commercetools-sync-java/commit/da4d6e9b71a4a85f83e34b2b996a98a74261cce2

File: src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+        final Optional<CustomObject<JsonNode>> fetchedCustomObjectOptional =
+                executeBlocking(customObjectService.fetchCustomObject(
+                        CustomObjectCompositeIdentifier.of(StringUtils.EMPTY,OLD_CUSTOM_OBJECT_CONTAINER)));
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java]: The method no longer exist in IT.

File: src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+                        CustomObjectCompositeIdentifier.of(OLD_CUSTOM_OBJECT_KEY, StringUtils.EMPTY)));
+        assertThat(fetchedCustomObjectOptional).isEmpty();
+    }
+
+
+    @Test
+    void fetchCustomObject_WithBlankKeyAndContainer_ShouldNotFetchCustomObject() {
+        final Optional<CustomObject<JsonNode>> fetchedCustomObjectOptional =
+            executeBlocking(customObjectService.fetchCustomObject(
+                    CustomObjectCompositeIdentifier.of(StringUtils.EMPTY,StringUtils.EMPTY)));
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/integration-test/java/com/commercetools/sync/integration/services/impl/CustomObjectServiceImplIT.java]: The method no longer exist in IT.

File: src/main/java/com/commercetools/sync/services/CustomObjectService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/services/CustomObjectService.java:34]: https://github.com/commercetools/commercetools-sync-java/pull/557/commits/b7931261ebfe6c704131a6eb4c6ce6ab10f357fe

File: src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java [hunk 17]
```diff
@@ -55,10 +62,22 @@ public CustomObjectServiceImpl(@Nonnull final CustomObjectSyncOptions syncOption
...
             @Nonnull final Set<CustomObjectCompositeIdentifier> identifiers) {
 
+        Set<CustomObjectCompositeIdentifier> filteredIdentifiers = identifiers.stream()
+                .filter(identifier ->
+                            StringUtils.isNotEmpty(identifier.getContainer())
+                        &&  StringUtils.isNotEmpty(identifier.getKey()))
+                .collect(Collectors.toSet());
+
+        if (filteredIdentifiers.size() == 0) {
+            return CompletableFuture.completedFuture(Collections.EMPTY_SET);
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java:72]:
```suggestion
            return CompletableFuture.completedFuture(Collections.emptySet());
```

File: src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+            String key = identifier.getKey();
+            String container = identifier.getContainer();
+            if (firstAttempt) {
+                queryPredicate = queryModel.container().is(container).and(queryModel.key().is(key));
+                firstAttempt = false;
+            } else {
+                queryPredicate = queryPredicate.or(queryModel.container().is(container).and(queryModel.key().is(key)));
+            }
+        }
+        return queryPredicate;
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java:104]: This method can be reduced to one reduce. Check if it works:

```suggestion
        return identifiers.stream()
                          .reduce(QueryPredicate.of(null),
                                  (partialQuery, identifier) -> {
                                      String key = identifier.getKey();
                                      String container = identifier.getContainer();
                                      return queryModel.container().is(container).and(queryModel.key().is(key));
                                  }, QueryPredicate::or);
```

File: src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java (hunk 17 above)
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java:72]: What's difference here?

File: src/main/java/com/commercetools/sync/customobjects/helpers/CustomObjectCompositeIdentifier.java [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    @Nonnull
+    public static CustomObjectCompositeIdentifier of(@Nonnull final CustomObject customObject) {
+        return new CustomObjectCompositeIdentifier(customObject.getKey(), customObject.getContainer());
+    }
+
+    /**
+     * Given a {@link String} and {@link String}, creates a {@link CustomObjectCompositeIdentifier} using the following
+     * fields from the supplied attributes.
+     * <ol>
+     * <li>{@link CustomObjectCompositeIdentifier#key}: key of {@link CustomObject#getKey()} </li>
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/customobjects/helpers/CustomObjectCompositeIdentifier.java]: https://github.com/commercetools/commercetools-sync-java/pull/557/commits/3abba1cc82361a5ca0d0e7c0cee392c361bfad06

File: src/main/java/com/commercetools/sync/services/CustomObjectService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/services/CustomObjectService.java]: https://github.com/commercetools/commercetools-sync-java/pull/557/commits/3abba1cc82361a5ca0d0e7c0cee392c361bfad06

File: src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java (hunk 17 above)
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java:72]: One is generified, the other one just returns Object as an generic type:
```
final CompletableFuture<Set> setCompletableFuture = CompletableFuture.completedFuture(Collections.EMPTY_SET);
final CompletableFuture<Set<CustomObject<JsonNode>>> setCompletableFuture = CompletableFuture.completedFuture(Collections.emptySet());
```

From the docs of `emptySet()`:
```
* Returns an empty set (immutable).  This set is serializable.
     * Unlike the like-named field, this method is parameterized.
     *
     * <p>This example illustrates the type-safe way to obtain an empty set:
     * <pre>
     *     Set&lt;String&gt; s = Collections.emptySet();
     * </pre>
```

File: src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java (hunk 18 above)
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java:104]: I added one integration test to test it but failed. So I keep the original one.
https://github.com/commercetools/commercetools-sync-java/pull/557/commits/9822feafbd5d2be063b25bbfea83ae2684383ff0

File: src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java (hunk 17 above)
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java]: Done
https://github.com/commercetools/commercetools-sync-java/pull/557/commits/5d32468abc9b76bc3de823da8b56751be997c019

**Review by lojzatran — APPROVED** [review]

**Review by JudeNiroshan — CHANGES_REQUESTED** [review]

File: src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+        }
+
+        Set<String> identifierStrings =
+                identifiers.stream().map(CustomObjectCompositeIdentifier::toString).collect(Collectors.toSet());
+
+        return fetchMatchingResources(identifierStrings,
+            draft -> CustomObjectCompositeIdentifier.of(draft).toString(),
+            () -> {
+                CustomObjectQueryBuilder<JsonNode> query = CustomObjectQueryBuilder.ofJsonNode();
+                query = query.plusPredicates(queryModel -> createQuery(queryModel, identifiers));
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java]: https://github.com/commercetools/commercetools-sync-java/pull/557/commits/af977b8acadfe59b7c94eb063470af40b33efbca

File: src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+                            StringUtils.isNotEmpty(identifier.getContainer())
+                        &&  StringUtils.isNotEmpty(identifier.getKey()))
+                .collect(Collectors.toSet());
+
+        if (filteredIdentifiers.size() == 0) {
+            return CompletableFuture.completedFuture(Collections.EMPTY_SET);
+        }
+
+        Set<String> identifierStrings =
+                identifiers.stream().map(CustomObjectCompositeIdentifier::toString).collect(Collectors.toSet());
```
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/main/java/com/commercetools/sync/services/impl/CustomObjectServiceImpl.java]: https://github.com/commercetools/commercetools-sync-java/pull/557/commits/af977b8acadfe59b7c94eb063470af40b33efbca

**Review by JudeNiroshan — APPROVED** [review]

**Comment by leungkinghin** [comment]
@lojzatran @JudeNiroshan @ninalindl 
Thank you for the help :)
